<center><img src='https://drive.google.com/uc?id=1_utx_ZGclmCwNttSe40kYA6VHzNocdET' height="60"></center>

AI TECH - Akademia Innowacyjnych Zastosowań Technologii Cyfrowych. Program Operacyjny Polska Cyfrowa na lata 2014-2020
<hr>

<center><img src='https://drive.google.com/uc?id=1BXZ0u3562N_MqCLcekI-Ens77Kk4LpPm'></center>

<center>
Projekt współfinansowany ze środków Unii Europejskiej w ramach Europejskiego Funduszu Rozwoju Regionalnego
Program Operacyjny Polska Cyfrowa na lata 2014-2020,
Oś Priorytetowa nr 3 "Cyfrowe kompetencje społeczeństwa" Działanie  nr 3.2 "Innowacyjne rozwiązania na rzecz aktywizacji cyfrowej"
Tytuł projektu:  „Akademia Innowacyjnych Zastosowań Technologii Cyfrowych (AI Tech)”
    </center>

# Statistical machine learning - Notebook 1

**Author: Maciej Bartczak**



# Bootcamp - introduction to machine learning: NumPy & Pandas


## What is NumPy?

NumPy is the fundamental package for scientific computing in Python. It is a Python library that provides a multidimensional array object and an assortment of routines for fast operations on arrays, including mathematical, logical, shape manipulation, sorting, selecting, I/O, basic linear algebra, statistical operations, random simulation and much more.

At the core of the NumPy package, is the ndarray object. This encapsulates n-dimensional arrays of homogeneous data types, with many operations being performed in compiled code for performance. There are several important differences between NumPy arrays and the standard Python sequences:

- The elements in a NumPy array are all required to be of the same data type, and thus will be the same size in memory. The exception: one can have arrays of (Python, including NumPy) objects, thereby allowing for arrays of different sized elements.

- NumPy arrays facilitate advanced mathematical and other types of operations on large numbers of data. Typically, **such operations are executed more efficiently and with less code than is possible using Python’s built-in sequences**.

- A growing plethora of scientific and mathematical Python-based packages are using NumPy arrays; though these typically support Python-sequence input, they convert such input to NumPy arrays prior to processing, and they often output NumPy arrays. In other words, in order to efficiently use much (perhaps even most) of today’s scientific/mathematical Python-based software, just knowing how to use Python’s built-in sequence types is insufficient - one also needs to know how to use NumPy arrays.



from: https://numpy.org/doc/stable/user/whatisnumpy.html

In [4]:
!pip install numpy matplotlib
import numpy as np
import matplotlib.pyplot as plt

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 8.7/8.7 MB 2.7 MB/s  0:00:03 eta 0:00:010m
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.9/4.9 MB 8.5 MB/s  0:00:00 eta 0:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.5/1.5 MB 9.1 MB/s  0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.0/7.0 MB 8.3 MB/s  0:00:00 eta 0:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7/7 [matplotlib]7 [matplotlib]


Matplotlib is building the font cache; this may take a moment.


## Array creation

There are multiple ways to create an array, try examples below.

### From nested lists

In [5]:
array_1d = np.array([1, 2, 3, 4, 5])
array_1d

array([1, 2, 3, 4, 5])

In [6]:
array_2d = np.array(
  [
    [1,  2,  3,  4,  5 ],
    [6,  7,  8,  9,  10],
    [11, 12, 13, 14, 15]
  ]
)
array_2d

array([[ 1,  2,  3,  4,  5],
       [ 6,  7,  8,  9, 10],
       [11, 12, 13, 14, 15]])

In [7]:
array_3d = np.array(
  [
    [
      [1,  2,  3,  4,  5 ],
      [6,  7,  8,  9,  10],
      [11, 12, 13, 14, 15],
    ],
    [
      [16, 17, 18, 19, 20],
      [21, 22, 23, 24, 25],
      [26, 27, 28, 29, 30],
    ]
  ]
)
array_3d

array([[[ 1,  2,  3,  4,  5],
        [ 6,  7,  8,  9, 10],
        [11, 12, 13, 14, 15]],

       [[16, 17, 18, 19, 20],
        [21, 22, 23, 24, 25],
        [26, 27, 28, 29, 30]]])

### Using zeros / ones / full



In [8]:
shape = (2, 3)

np.zeros(shape)

array([[0., 0., 0.],
       [0., 0., 0.]])

In [9]:
np.ones(shape)

array([[1., 1., 1.],
       [1., 1., 1.]])

In [10]:
fill_value = 42
np.full(shape, fill_value)

array([[42, 42, 42],
       [42, 42, 42]])

### Using zeros_like / ones_like / full_like

In [11]:
np.zeros_like(array_2d)

array([[0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0]])

In [12]:
np.ones_like(array_2d)

array([[1, 1, 1, 1, 1],
       [1, 1, 1, 1, 1],
       [1, 1, 1, 1, 1]])

In [13]:
fill_value = 42
np.full_like(array_2d, fill_value)

array([[42, 42, 42, 42, 42],
       [42, 42, 42, 42, 42],
       [42, 42, 42, 42, 42]])

### Using arange / linspace (1D arrays only)

which are used to fill an array with values from the interval $[low, high]$ with specific step or number of points.

In [14]:
start = 0
stop  = 1
step  = 0.1

np.arange(start, stop, step)

array([0. , 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9])

In [15]:
np.arange(start, stop + 1e-4, step)

array([0. , 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1. ])

In [16]:
n_steps = int((stop-start)/step) + 1
np.linspace(start, stop, n_steps)

array([0. , 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1. ])

### Using identity / diag (2D arrays only)
 to create identity or diagonal (with optional offset) matrices.

In [17]:
n = 4
np.identity(n)

array([[1., 0., 0., 0.],
       [0., 1., 0., 0.],
       [0., 0., 1., 0.],
       [0., 0., 0., 1.]])

In [18]:
np.diag([1, 2, 3, 4])

array([[1, 0, 0, 0],
       [0, 2, 0, 0],
       [0, 0, 3, 0],
       [0, 0, 0, 4]])

In [19]:
offset = 1
np.diag([1, 2, 3], offset)

array([[0, 1, 0, 0],
       [0, 0, 2, 0],
       [0, 0, 0, 3],
       [0, 0, 0, 0]])

### Using meshgrid
For 1D arrays X_1, ..., X_n meshgrid behavior is specified as follows

```
G_1, ..., G_n = np.meshgrid(X_1, ..., X_n, indexing='ij')
G_j[i_1, i_2, ..., i_n] = X_j[i_j]

G_1, ..., G_n = np.meshgrid(X_1, ..., X_n, indexing='xy')  # default indexing
G_j[i_2, i_1, i_3, ..., i_n] = X_j[i_j]
```

what facilitates creating coordinate grids,




In [20]:
x = np.array([1, 2, 3], dtype=int)
y = np.array([4, 5, 6, 7], dtype=int)

x_grid, y_grid = np.meshgrid(x, y, indexing='ij')

such as:

In [21]:
x_grid

array([[1, 1, 1, 1],
       [2, 2, 2, 2],
       [3, 3, 3, 3]])

In [22]:
y_grid

array([[4, 5, 6, 7],
       [4, 5, 6, 7],
       [4, 5, 6, 7]])

In [23]:
x_grid.shape == y_grid.shape == (len(x), len(y))

True

### Using tile & repeat
to create an array filled with "repeating pattern".

In [24]:
array = np.array([0,1,2])

In [25]:
np.tile(array, 2)

array([0, 1, 2, 0, 1, 2])

In [26]:
np.tile(array, (2, 1))

array([[0, 1, 2],
       [0, 1, 2]])

In [27]:
np.repeat(array, 2)

array([0, 0, 1, 1, 2, 2])

In [28]:
np.repeat(array, 2, axis=0)

array([0, 0, 1, 1, 2, 2])

In [29]:
np.repeat(array.reshape(1, -1), 2, axis=0)

array([[0, 1, 2],
       [0, 1, 2]])

In [30]:
np.repeat(array.reshape(1, -1), 2, axis=1)

array([[0, 0, 1, 1, 2, 2]])

## NumPy dtypes

Full list of NumPy dtypes is available here: https://numpy.org/doc/stable/user/basics.types.html

Among which there are 5 basic numerical types representing booleans (bool), integers (int), unsigned integers (uint) floating point (float) and complex.

Data-types can be used as arguments to the dtype keyword that many numpy functions or methods accept, in particular array creation routines.

See examples below.

In [31]:
np.array([1, 2, 3, 4, 5], dtype=int)

array([1, 2, 3, 4, 5])

In [32]:
np.array([1, 2, 3, 4, 5], dtype=float)

array([1., 2., 3., 4., 5.])

In [33]:
np.array([1, 2, 3, 4, 5], dtype=complex)

array([1.+0.j, 2.+0.j, 3.+0.j, 4.+0.j, 5.+0.j])

Types can be cast as follows

In [34]:
float_array   = np.array([1, 2, 3, 4, 5], dtype=float)
complex_array = float_array.astype(complex)

float_array, float_array.dtype

(array([1., 2., 3., 4., 5.]), dtype('float64'))

In [35]:
complex_array, complex_array.dtype

(array([1.+0.j, 2.+0.j, 3.+0.j, 4.+0.j, 5.+0.j]), dtype('complex128'))

## Saving / loading data

### Using savetxt / loadtxt

to process a single 1D/2D array in human readable format. File created with `savetxt` can be easily accessed with a text editor

In [36]:
data = np.array(
  [
    [1, 2],
    [3, 4],
  ]
)

np.savetxt('data.txt', data)
!cat data.txt

1.000000000000000000e+00 2.000000000000000000e+00
3.000000000000000000e+00 4.000000000000000000e+00


as well as seamlessly loaded back with `loadtxt`.


In [37]:
np.loadtxt('data.txt')

array([[1., 2.],
       [3., 4.]])

File format can be specified, see https://numpy.org/doc/stable/reference/generated/numpy.savetxt.html for reference

In [38]:
np.savetxt('data.txt', data, fmt='%.2f')
!cat data.txt

1.00 2.00
3.00 4.00


In [39]:
np.savetxt('data.txt', data, fmt='%d')
!cat data.txt

1 2
3 4


Nevertheless information regarding dtype is not preserved.

In [40]:
np.loadtxt('data.txt')

array([[1., 2.],
       [3., 4.]])

In [41]:
np.loadtxt('data.txt', dtype=int)

array([[1, 2],
       [3, 4]])

`savetxt` method can be used to save an array in csv format, we only need to specify the proper delimiter

In [42]:
np.savetxt('data.csv', data, delimiter=',', fmt='%.2f')
!cat data.csv

1.00,2.00
3.00,4.00


as well as loaded back

In [43]:
np.loadtxt('data.csv', delimiter=',')

array([[1., 2.],
       [3., 4.]])

### Using save / load
to process an arbitrary array in binary format. Unfortunately the resulting file is not human readable. On the other hand these methods can handle array of arbitrary shape and preserve the underlying dtype.

In [44]:
data = np.array(
  [
    [
      [1, 2],
      [3, 4],
    ]
  ]
)

np.save('data.npy', data)

!cat data.npy

�NUMPYv{'descr': '<i8', 'fortran_order': False, 'shape': (1, 2, 2), }                                                       


In [45]:
np.load('data.npy')

array([[[1, 2],
        [3, 4]]])

In [46]:
np.load('data.npy').dtype == data.dtype

True

### Using savez / load

to process **multiple** arbitrary arrays in binary format. The result of the load operation is an object of dictionary-like structure.

In [47]:
array_1 = np.array([1, 2, 3], dtype=int)
array_2 = np.array([[4, 5, 6]], dtype=float)
array_3 = np.array([[[7, 8, 9]]], dtype=complex)

np.savez('data.npz', array_1, array_2, some_name=array_3)
loaded_data = np.load('data.npz')
loaded_data.files

['some_name', 'arr_0', 'arr_1']

In [48]:
loaded_data['arr_0']

array([1, 2, 3])

In [49]:
loaded_data['arr_1']

array([[4., 5., 6.]])

In [50]:
loaded_data['some_name']

array([[[7.+0.j, 8.+0.j, 9.+0.j]]])

### Using savez_compressed / load
to process multiple arbitrary arrays in binary format with compression. It behaves just as savez but the resulting file is also compressed, what save space on the disk.



In [51]:
shape = (1000, 1000)
data = np.ones(shape)
np.savez('data_uncompressed.npz', data)
np.savez_compressed('data_compressed.npz', data)

We can compare the filesizes with `ls` command.

In [52]:
!ls data_*compressed.npz -lh

-rw-rw-r-- 1 sysia sysia  12K Oct 20 22:21 data_compressed.npz
-rw-rw-r-- 1 sysia sysia 7.7M Oct 20 22:21 data_uncompressed.npz


## Shape manipulation
It is possible to change the structure of data in the array. See examples below.

#### Reshape

In [53]:
data = np.arange(0, 12, 1, dtype = int)
data

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11])

In [54]:
data.reshape(3, 4)

array([[ 0,  1,  2,  3],
       [ 4,  5,  6,  7],
       [ 8,  9, 10, 11]])

In [55]:
data.reshape(4, 3)

array([[ 0,  1,  2],
       [ 3,  4,  5],
       [ 6,  7,  8],
       [ 9, 10, 11]])

In [56]:
data_reshaped = data.reshape(2, 3, 2)
data_reshaped

array([[[ 0,  1],
        [ 2,  3],
        [ 4,  5]],

       [[ 6,  7],
        [ 8,  9],
        [10, 11]]])

In [57]:
shape = data_reshaped.shape
shape

(2, 3, 2)

#### Inferring **one** dimension while usigng reshape method

In [58]:
data.reshape(6, -1)

array([[ 0,  1],
       [ 2,  3],
       [ 4,  5],
       [ 6,  7],
       [ 8,  9],
       [10, 11]])

In [59]:
data.reshape(-1, 6)

array([[ 0,  1,  2,  3,  4,  5],
       [ 6,  7,  8,  9, 10, 11]])

In [60]:
data.reshape(3, 2, -1)

array([[[ 0,  1],
        [ 2,  3]],

       [[ 4,  5],
        [ 6,  7]],

       [[ 8,  9],
        [10, 11]]])

#### Flatten

In [61]:
data_reshaped.flatten()

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11])

#### Adding / deleting dummy dimensions
Facilites vectorized arithmetic with conjunction with broadcasting mechanism (explained later)

In [62]:
data

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11])

In [63]:
data.shape

(12,)

In [64]:
data_expanded1 = np.expand_dims(data, 0)
data_expanded1

array([[ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11]])

In [65]:
data_expanded1.shape

(1, 12)

In [66]:
data_expanded2 = np.expand_dims(data, 1)
data_expanded2

array([[ 0],
       [ 1],
       [ 2],
       [ 3],
       [ 4],
       [ 5],
       [ 6],
       [ 7],
       [ 8],
       [ 9],
       [10],
       [11]])

In [67]:
data_expanded2.shape

(12, 1)

In [68]:
data_expanded3 = np.expand_dims(data_expanded1, 2)
data_expanded3

array([[[ 0],
        [ 1],
        [ 2],
        [ 3],
        [ 4],
        [ 5],
        [ 6],
        [ 7],
        [ 8],
        [ 9],
        [10],
        [11]]])

In [69]:
data_expanded3.shape

(1, 12, 1)

In [70]:
data_squeezed = data_expanded3.squeeze()
data_squeezed

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11])

In [71]:
data_squeezed.shape

(12,)

#### Transposing

For 2D arrays it behaves just as expected (performs matrix transpose).

In [72]:
data = np.arange(0, 4, 1).reshape(2, 2)
data

array([[0, 1],
       [2, 3]])

In [73]:
data.transpose()

array([[0, 2],
       [1, 3]])

For higher dimensional arrays its best visualized via observing how the shape changes.

In [74]:
shape = (1, 2, 3, 4)
data = np.ones(shape)
data.shape

(1, 2, 3, 4)

The default behavior is to flip dimensions.

In [75]:
data.transpose().shape

(4, 3, 2, 1)

However other permutations of dimensions might be specified.

In [76]:
dimensions_permutation = [1, 0, 3, 2]
data.transpose(dimensions_permutation).shape

(2, 1, 4, 3)

`.T` is a shortand for `.transpose()`

In [77]:
data.T.shape

(4, 3, 2, 1)

## Concatenation & stacking
The following methods are used to merge multiple arrays (with compatible shapes) into one array.

### concatenate
used to join a sequence of arrays along an **existing** axis.

In [78]:
shape = (4,)
array_1 = np.full(shape, 1)
array_2 = np.full(shape, 2)

array_1, array_2

(array([1, 1, 1, 1]), array([2, 2, 2, 2]))

In [79]:
np.concatenate([array_1, array_2, array_1])

array([1, 1, 1, 1, 2, 2, 2, 2, 1, 1, 1, 1])

In [80]:
array_1 = array_1.reshape(1, -1)
array_2 = array_2.reshape(1, -1)

array_1, array_2

(array([[1, 1, 1, 1]]), array([[2, 2, 2, 2]]))

The default axis to join on is 0

In [81]:
np.concatenate([array_1, array_2, array_1])

array([[1, 1, 1, 1],
       [2, 2, 2, 2],
       [1, 1, 1, 1]])

But it can be changed

In [82]:
np.concatenate([array_1, array_2, array_1], axis=1)

array([[1, 1, 1, 1, 2, 2, 2, 2, 1, 1, 1, 1]])

### stack

used to stack a sequence of arrays along a **new** axis.

In [83]:
shape = (4,)
array_1 = np.full(shape, 1)
array_2 = np.full(shape, 2)

array_1, array_2

(array([1, 1, 1, 1]), array([2, 2, 2, 2]))

In [84]:
np.stack([array_1, array_2, array_1])

array([[1, 1, 1, 1],
       [2, 2, 2, 2],
       [1, 1, 1, 1]])

In [85]:
np.stack([array_1, array_2, array_1], axis=1)

array([[1, 2, 1],
       [1, 2, 1],
       [1, 2, 1],
       [1, 2, 1]])


### hstack, vstack and dstack
stand for horizontal, vertical and depthwise stacking

In [86]:
np.hstack([array_1, array_2, array_1])

array([1, 1, 1, 1, 2, 2, 2, 2, 1, 1, 1, 1])

In [87]:
np.vstack([array_1, array_2, array_1])

array([[1, 1, 1, 1],
       [2, 2, 2, 2],
       [1, 1, 1, 1]])

In [88]:
np.dstack([array_1, array_2, array_1])

array([[[1, 2, 1],
        [1, 2, 1],
        [1, 2, 1],
        [1, 2, 1]]])

## Mathematical operations
As stated before

> NumPy arrays facilitate advanced mathematical and other types of operations on large numbers of data. Typically, such operations are executed more efficiently and with less code than is possible using Python’s built-in sequences.

Let us observe some examples.



In [89]:
x = np.arange(0, 3, 1)
y = np.arange(3, 6, 1)

x, y

(array([0, 1, 2]), array([3, 4, 5]))

#### Elementwise

Basic mathematical operations will be applied to the arrays elementwise.

In [90]:
2 * x

array([0, 2, 4])

In [91]:
x * 2

array([0, 2, 4])

In [92]:
np.sin(x)

array([0.        , 0.84147098, 0.90929743])

In [93]:
x**2

array([0, 1, 4])

The same applies to the binary (taking two arguments) operators and arrays of the same shape.

In [94]:
x + y

array([3, 5, 7])

In [95]:
x * y

array([ 0,  4, 10])

In [96]:
x / y

array([0.  , 0.25, 0.4 ])

In [97]:
y ** x

array([ 1,  4, 25])

#### Broadcasting

When arrays have different shapes the broadcasting mechanism comes into play.

The term broadcasting describes how NumPy treats arrays with different shapes during arithmetic operations. Subject to certain constraints, the smaller array is “broadcast” across the larger array so that they have compatible shapes. Broadcasting provides a means of vectorizing array operations so that looping occurs in C instead of Python. It does this without making needless copies of data and usually leads to efficient algorithm implementations.

When operating on two arrays, NumPy compares their shapes element-wise. It starts with the trailing (i.e. rightmost) dimensions and works its way left. Two dimensions are compatible when

- they are equal, or
- one of them is 1

If these conditions are not met, a `ValueError: operands could not be broadcast together` exception is thrown, indicating that the arrays have incompatible shapes. The size of the resulting array is the size that is not 1 along each axis of the inputs.

from: https://numpy.org/devdocs/user/basics.broadcasting.html

In [98]:
x_shape = (2, 4)
y_shape = (2, 1)

x = np.ones(x_shape)
y = np.ones(y_shape)

(x + y).shape

(2, 4)

In [99]:
x_shape =    (2, 4)
y_shape = (3, 1, 1)

x = np.ones(x_shape)
y = np.ones(y_shape)

(x + y).shape

(3, 2, 4)

In [100]:
x_shape = (   1, 4)
y_shape = (3, 2, 1)

x = np.ones(x_shape)
y = np.ones(y_shape)

(x + y).shape

(3, 2, 4)

### Matrix / tensor operations
NumPy package wouldn't be complete without means to manipulate arrays other than elementwise operations, such as most common linear algebra functionalities. See the examples below.

#### Matrix multiplication

can be performed with `matmul` method

In [101]:
a = np.eye(3)
a[2,2] = 2
b = np.arange(0, 3, 1).reshape(3, 1)

a

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 2.]])

In [102]:
b

array([[0],
       [1],
       [2]])

In [103]:
np.matmul(a, b)

array([[0.],
       [1.],
       [4.]])

or `@` operand

In [104]:
a @ b

array([[0.],
       [1.],
       [4.]])

Be mindfull of its varying behaviour for different input shapes

- If both arguments are 2-D they are multiplied like conventional matrices.
- If either argument is N-D, N > 2, it is treated as a stack of matrices residing in the last two indexes and broadcast accordingly.
- If the first argument is 1-D, it is promoted to a matrix by prepending a 1 to its dimensions. After matrix multiplication the prepended 1 is removed.
- If the second argument is 1-D, it is promoted to a matrix by appending a 1 to its dimensions. After matrix multiplication the appended 1 is removed.

from: https://numpy.org/doc/stable/reference/generated/numpy.matmul.html

#### Dot product

can be performed using `dot` method.

In [105]:
x = np.arange(0, 3, 1)
y = np.arange(3, 6, 1)

x, y

(array([0, 1, 2]), array([3, 4, 5]))

In [106]:
np.dot(x, y)

np.int64(14)

Be mindfull of its varying behaviour for different input shapes

- If both a and b are 1-D arrays, it is inner product of vectors (without complex conjugation).

- If both a and b are 2-D arrays, it is matrix multiplication, but using matmul or a @ b is preferred.

- If either a or b is 0-D (scalar), it is equivalent to multiply and using numpy.multiply(a, b) or a * b is preferred.

- If a is an N-D array and b is a 1-D array, it is a sum product over the last axis of a and b.

- If a is an N-D array and b is an M-D array (where M>=2), it is a sum product over the last axis of a and the second-to-last axis of b:

  ```
  dot(a, b)[i,j,k,m] = sum(a[i,j,:] * b[k,:,m])
  ```

from: https://numpy.org/doc/stable/reference/generated/numpy.dot.html


#### Einstein summation convention
Einstein summation convention provide a concise way of expressing multilinear operations. You can look it up here https://mathworld.wolfram.com/EinsteinSummation.html and its NumPy specification here https://numpy.org/doc/stable/reference/generated/numpy.einsum.html.

Take a look a the following ilustrative examples.


In [107]:
data = np.tri(3)
data

array([[1., 0., 0.],
       [1., 1., 0.],
       [1., 1., 1.]])

In [108]:
np.einsum('ij->ji', data)  # transpose

array([[1., 1., 1.],
       [0., 1., 1.],
       [0., 0., 1.]])

In [109]:
np.einsum('ij->i', data)  # row sum

array([1., 2., 3.])

In [110]:
np.einsum('ij->j', data)  # column sum

array([3., 2., 1.])

In [111]:
np.einsum('ii->i', data)  # diagonal

array([1., 1., 1.])

In [112]:
vector = np.ones(3)
vector

array([1., 1., 1.])

In [113]:
np.einsum('ij,j->i', data, vector)  # matrix multiplication (matrix x vector)

array([1., 2., 3.])

In [114]:
np.einsum('ij,jk->ik', data, vector.reshape(3, 1))  # matrix multiplication (matrix x matrix)

array([[1.],
       [2.],
       [3.]])

## Other operations

### Aggregations: mean / var
Very often we are required to calculate some aggregations with respect to specific dimensions of an array, such as mean or variance.

In [115]:
data = np.arange(0, 20, 1).reshape(4, 5)
data

array([[ 0,  1,  2,  3,  4],
       [ 5,  6,  7,  8,  9],
       [10, 11, 12, 13, 14],
       [15, 16, 17, 18, 19]])

By default the whole array is aggregated

In [116]:
data.mean()

np.float64(9.5)

But aggregations can also be applied axiswise

In [117]:
data.mean(axis=0)

array([ 7.5,  8.5,  9.5, 10.5, 11.5])

In [118]:
data.mean(axis=1)

array([ 2.,  7., 12., 17.])

In [119]:
data.var(axis=0)

array([31.25, 31.25, 31.25, 31.25, 31.25])

In [120]:
data.var(axis=1)

array([2., 2., 2., 2.])

### Aggregation: quantiles

Many more aggregations are possible, e.g. determining data quantiles.

In [121]:
data = np.linspace(1, 2, 101)

np.quantile(data, 0.3)

np.float64(1.3)

In [122]:
np.quantile(data, [0.3, 0.6])

array([1.3, 1.6])

### Sorting
Arrays can be sorted. By default sorting is performed with respect to the last dimension

In [123]:
data = np.array([
  [1, 4, 2, 8, 5],
  [2, 3, 1, 9, 6],
  [8, 2, 3, 7, 4],
])

np.sort(data)

array([[1, 2, 4, 5, 8],
       [1, 2, 3, 6, 9],
       [2, 3, 4, 7, 8]])

but sorting dimension can be explicitly specified.

In [124]:
np.sort(data, axis=0)

array([[1, 2, 1, 7, 4],
       [2, 3, 2, 8, 5],
       [8, 4, 3, 9, 6]])

In [125]:
np.sort(data, axis=1)

array([[1, 2, 4, 5, 8],
       [1, 2, 3, 6, 9],
       [2, 3, 4, 7, 8]])

You can also get the "sorting permutations" that can be used in conjunction with advanced indexing (introduced below).

In [126]:
np.argsort(data, axis=0)

array([[0, 2, 1, 2, 2],
       [1, 1, 0, 0, 0],
       [2, 0, 2, 1, 1]])

In [127]:
np.argsort(data, axis=1)

array([[0, 2, 1, 4, 3],
       [2, 0, 1, 4, 3],
       [1, 2, 4, 3, 0]])

## Indexing

Allows selecting items and subarrays. It can be also used to modify fragments of an array.

### Basic indexing

In [128]:
data = np.arange(0, 16, 1)
data

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15])

In [129]:
data[10]

np.int64(10)

In [130]:
data[-2]

np.int64(14)

In [131]:
data[:5]

array([0, 1, 2, 3, 4])

In [132]:
data[::2]

array([ 0,  2,  4,  6,  8, 10, 12, 14])

In [133]:
data[-3:]

array([13, 14, 15])

It is also possible in higher dimensional setting


In [134]:
data = data.reshape(4, 4)
data

array([[ 0,  1,  2,  3],
       [ 4,  5,  6,  7],
       [ 8,  9, 10, 11],
       [12, 13, 14, 15]])

In [135]:
data[1, 1]

np.int64(5)

In [136]:
data[1, :]

array([4, 5, 6, 7])

One does not need to specify trailing colons `:`

In [137]:
data[:-2]

array([[0, 1, 2, 3],
       [4, 5, 6, 7]])

Note how the column selection is achieved. In this case `:` can't be ommited.

In [138]:
data[:, -2:]

array([[ 2,  3],
       [ 6,  7],
       [10, 11],
       [14, 15]])

One group of subsequent colons `:` can be replaced with `...`

In [139]:
data_reshaped = data.reshape(4, 2, 2)
data_reshaped

array([[[ 0,  1],
        [ 2,  3]],

       [[ 4,  5],
        [ 6,  7]],

       [[ 8,  9],
        [10, 11]],

       [[12, 13],
        [14, 15]]])

In [140]:
data_reshaped[..., 0]

array([[ 0,  2],
       [ 4,  6],
       [ 8, 10],
       [12, 14]])

Note the difference

In [141]:
data_reshaped[:, 0]

array([[ 0,  1],
       [ 4,  5],
       [ 8,  9],
       [12, 13]])

### View & copies

Basic indexing routines presented above return views of the data in the underlying arrays. When the underlying array is modified the changes are reflected in the view.

In [142]:
data = np.arange(0, 16, 1).reshape(4, 4)
data

array([[ 0,  1,  2,  3],
       [ 4,  5,  6,  7],
       [ 8,  9, 10, 11],
       [12, 13, 14, 15]])

In [143]:
view = data[:, 3]
view

array([ 3,  7, 11, 15])

In [144]:
data *= 2
data

array([[ 0,  2,  4,  6],
       [ 8, 10, 12, 14],
       [16, 18, 20, 22],
       [24, 26, 28, 30]])

In [145]:
view

array([ 6, 14, 22, 30])

To "detach" view from the underlying array use `copy` method

In [146]:
view_copy = view.copy()
data *= 2
view, view_copy

(array([12, 28, 44, 60]), array([ 6, 14, 22, 30]))

Views are only possible when selecting with slices, then it's a matter of rememebering offset and stride only.

You can read more about this here: https://scipy-cookbook.readthedocs.io/items/ViewsVsCopies.html

### Advanced indexing

it is also possible to index with integer or boolean arrays

In [147]:
data = np.arange(0, 16, 1).reshape(4, 4)
data

array([[ 0,  1,  2,  3],
       [ 4,  5,  6,  7],
       [ 8,  9, 10, 11],
       [12, 13, 14, 15]])

We can select specific rows

In [148]:
data[[0, 3]]

array([[ 0,  1,  2,  3],
       [12, 13, 14, 15]])

and columns.

In [149]:
data[:, [0, 3]]

array([[ 0,  3],
       [ 4,  7],
       [ 8, 11],
       [12, 15]])

Seleceting both together is a little more complicated

In [150]:
rows = [
 [0, 0],
 [3, 3]
]
columns = [
 [0, 3],
 [0, 3]
]
data[rows, columns]

array([[ 0,  3],
       [12, 15]])

Use indices shape to control the output shape.

In [151]:
rows = [0, 0, 3, 3]
columns = [0, 3, 0, 3]
data[rows, columns]

array([ 0,  3, 12, 15])

Boolean arrays can be used as well, though they work somewhat differently.

In [152]:
index = (data % 3 == 0)
index

array([[ True, False, False,  True],
       [False, False,  True, False],
       [False,  True, False, False],
       [ True, False, False,  True]])

In [153]:
data[index]

array([ 0,  3,  6,  9, 12, 15])

Its most common use case is to update array entries based on some condition.

In [154]:
data[index] = -1
data

array([[-1,  1,  2, -1],
       [ 4,  5, -1,  7],
       [ 8, -1, 10, 11],
       [-1, 13, 14, -1]])

Detailed bahavior of indexing with boolean arrays is described here: https://numpy.org/doc/stable/reference/arrays.indexing.html#advanced-indexing

# NumPy Exercises

### Exercise: multiplication table
Create "{0, ..., 10} x {0, ..., 10} multiplication table" using only
- `meshgrid` method,
- multiplication of 1D arrays, dummy dimensions and broadcasting mechanism.

In [9]:
### YOUR CODE BEGINS HERE ###
import numpy as np
arr = np.arange(0, 11, 1)
arr2d = np.meshgrid(arr, arr, indexing='ij')[0]
arr = arr2d @ arr2d.T / 11
print(np.array(arr))

### YOUR CODE ENDS HERE ###

[[  0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.]
 [  0.   1.   2.   3.   4.   5.   6.   7.   8.   9.  10.]
 [  0.   2.   4.   6.   8.  10.  12.  14.  16.  18.  20.]
 [  0.   3.   6.   9.  12.  15.  18.  21.  24.  27.  30.]
 [  0.   4.   8.  12.  16.  20.  24.  28.  32.  36.  40.]
 [  0.   5.  10.  15.  20.  25.  30.  35.  40.  45.  50.]
 [  0.   6.  12.  18.  24.  30.  36.  42.  48.  54.  60.]
 [  0.   7.  14.  21.  28.  35.  42.  49.  56.  63.  70.]
 [  0.   8.  16.  24.  32.  40.  48.  56.  64.  72.  80.]
 [  0.   9.  18.  27.  36.  45.  54.  63.  72.  81.  90.]
 [  0.  10.  20.  30.  40.  50.  60.  70.  80.  90. 100.]]


### Exercise: speed comparison
In this exercise we will measure the speed improvements due to NumPy implementation.

In [157]:
length = 10000
a = b = np.arange(length)

Write the code to calculate dotproduct of arrays `a` and `b` using only native Python funcions. Use `timeit` magic to time the execution.

In [158]:
%%timeit

### YOUR CODE BEGINS HERE ###

### YOUR CODE ENDS HERE ###

ValueError: empty body on For

Now compare the execution time with `np.dot`.

In [ ]:
### YOUR CODE BEGINS HERE ###

### YOUR CODE ENDS HERE ###

### Exercise: vector operations
Calculate
$$ \int_0^1 \sin(a\cdot2\pi x)\sin(b\cdot 2\pi x)dx $$
for
- $a=b=1$,
- $a=1$ and $b=2$,

using discrete approximation of an integral and following NumPy functions: `linspace, sin, mean`. Do not use `vectorize`.


In [19]:
# Case a = b = 1
### YOUR CODE BEGINS HERE ###
parts = 3
a = b = 1
answer = np.dot(np.sin(2 * np.pi * a * np.arange(0, 1, 1/parts)), np.sin(2 * np.pi * b * np.arange(0, 1, 1/parts))) / parts

### YOUR CODE ENDS HERE ###

print(answer.round(4))
assert np.isclose(answer, 0.5)

0.5


In [ ]:
np.linspace(0, 1, 10, endpoint=False)
np.linspace(0, 1, 11)[:-1]

In [ ]:
# Case a = 1, b = 2
### YOUR CODE BEGINS HERE ###

### YOUR CODE ENDS HERE ###
print(answer.round(4))
assert np.isclose(answer, 0)

### Exercise: vector operations ctd.

Let $a_j = j$. Calculate in a similar fashion matrix $(b_{ij})_{i,j=0, ..., 4}$, where
$$ b_{ij} = \int_0^1 \sin(a_i\cdot2\pi x)\sin(a_j\cdot 2\pi x)dx. $$

Try to structure your code in such a way that it would accept an arbitrary array `a`. Do not use `vectorize`.

In [ ]:
a = np.arange(5)

### YOUR CODE BEGINS HERE ###

### YOUR CODE ENDS HERE ###

print(answer.round(2))
assert np.isclose(answer, np.diag([0, 0.5, 0.5, 0.5, 0.5])).all()

### Exercise: subarrays

Let us sample $n=100 000$ standard normal random variables. Calculate mean of values greater than zero.

In [ ]:
np.random.seed(0)

n = 100000
sample = np.random.randn(n)

### YOUR CODE BEGINS HERE ###

### YOUR CODE ENDS HERE ###

assert np.isclose(answer, np.sqrt(2/np.pi), rtol=0.01)

### Exercise: subarrays ctd.
Let us sample an array of random integers.
1. Select columns for which sum of its values gives remainder = 1 when divided by 3.
2. Change entries in that columns to ones.
3. Calculate sum of all entries.

In [ ]:
np.random.seed(0)
sample = np.random.randint(0, 3, size=(10, 10))

### YOUR CODE BEGINS HERE ###

### YOUR CODE ENDS HERE ###

assert answer == 88


### Exercise: matrix multiplication & merging arrays

In this exercies we will approximate operator norm of some matrices.

1. Create an an array of $n=1000$ vectors $v\in\mathbb{R}^2$ "evenly spaced" on the unit circle.
2. Combine given three matrices into one array.
3. Use **one** matrix multiplication operation with these arrays to obtain matrix products of all the vectors with all the matrices.
4. For each vector calculate its (euclidean) length.
5. Take maximum with respect to each matrix.
6. You should obtain three values representing approximations of operator norm of provided matrices.


In [ ]:
matrix_1 = np.array([
  [1, 0],
  [0, 1]
])

matrix_2 = np.array([
  [1, 1],
  [1, 1]
])

matrix_3 = np.array([
  [1, 1],
  [0, 1]
])

### YOUR CODE BEGINS HERE ###

### YOUR CODE ENDS HERE ###

print(answer.round(4))
assert np.isclose(answer, [1, 2, (1+np.sqrt(5))/2]).all()

## Other NumPy functionalities
NumPy package implements plethora of mathematical functionalities, not covered in this tutorial. NumPy User guide and API Reference are great starting point to discover those functionalities:
- https://numpy.org/doc/stable/user/
- https://numpy.org/doc/stable/reference/


# What is pandas?

Pandas is a fast, powerful, flexible and easy to use open source data analysis and manipulation tool,
built on top of the Python programming language.

It is meant for manipulating datasets in tabular format, that is 2D arrays with rows representing observations and columns representing variables, the so called **DataFrames**.

from: https://pandas.pydata.org/

### Creating dataframes

In the most basic form you just specify 2D array of values

In [ ]:
%pip install pandas
import pandas as pd
import numpy as np

df = pd.DataFrame(
    np.arange(0, 16, 1).reshape(4, 4)
)

df

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.4/12.4 MB 3.9 MB/s  0:00:03 eta 0:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2/2 [pandas]2m1/2 [pandas]


0   1   2   3
0   0   1   2   3
1   4   5   6   7
2   8   9  10  11
3  12  13  14  15

It is possible to additionally specify columns and index labels

In [22]:
df = pd.DataFrame(
    np.arange(0, 16, 1).reshape(4, 4),
    columns = ['column_1', 'column_2', 'column_3', 'column_4'],
    index   = ['a', 'b', 'c', 'd']
)

df

column_1  column_2  column_3  column_4
a         0         1         2         3
b         4         5         6         7
c         8         9        10        11
d        12        13        14        15

as well as type of entries

In [23]:
df = pd.DataFrame(
    np.arange(0, 16, 1).reshape(4, 4),
    columns = ['column_1', 'column_2', 'column_3', 'column_4'],
    index   = ['a', 'b', 'c', 'd'],
    dtype   = float
)

df

column_1  column_2  column_3  column_4
a       0.0       1.0       2.0       3.0
b       4.0       5.0       6.0       7.0
c       8.0       9.0      10.0      11.0
d      12.0      13.0      14.0      15.0

In [24]:
df.astype(int)

column_1  column_2  column_3  column_4
a         0         1         2         3
b         4         5         6         7
c         8         9        10        11
d        12        13        14        15

It can be done in an alternative manner

In [25]:
df = pd.DataFrame(
    {
      'column_1': [0,  4,  8, 12],
      'column_2': [1,  5,  9, 13],
      'column_3': [2,  6, 10, 14],
      'column_4': [3,  7, 11, 15]
    },
    index   = ['a', 'b', 'c', 'd']
)

df

column_1  column_2  column_3  column_4
a         0         1         2         3
b         4         5         6         7
c         8         9        10        11
d        12        13        14        15

It is worth mentioning that the underlying data can be accesed under `values` attribute

In [26]:
df.values

array([[ 0,  1,  2,  3],
       [ 4,  5,  6,  7],
       [ 8,  9, 10, 11],
       [12, 13, 14, 15]])

What's important Pandas DataFrames can handle entries of different type. In the most common situation different columns will have different types.

In [27]:
df = pd.DataFrame(
    {
      'integers': [0,   1,   2,  3],
      'floats':   [1.0, 2.0, 3,  4],
      'booleans': [True, False, True, False],
      'arbitrary objects':  [0, 1.0, 'abc', False]
    }
)

df

integers  floats  booleans arbitrary objects
0         0     1.0      True                 0
1         1     2.0     False               1.0
2         2     3.0      True               abc
3         3     4.0     False             False

In [28]:
df.dtypes

integers               int64
floats               float64
booleans                bool
arbitrary objects     object
dtype: object

Another available creation methods are listed below
- `pd.DataFrame.from_dict`
- `pd.DataFrame.from_records`

You can read about them here: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html

### Loading data
Let us peek into the CSV file conveniently provided in the Google Colab runtime

In [ ]:
!head './sample_data/california_housing_train.csv'

/home/sysia/Desktop/public-dnn-2025-26/docs
head: cannot open './sample_data/california_housing_train.csv' for reading: No such file or directory


*and* load it using `pd.read_csv` method.

In [30]:
pd.read_csv('./sample_data/california_housing_train.csv')

FileNotFoundError: [Errno 2] No such file or directory: './sample_data/california_housing_train.csv'

In Pandas it is also possible to load JSON files.

In [37]:
!head './sample_data/anscombe.json'  --- IGNORE --

head: cannot open './sample_data/anscombe.json' for reading: No such file or directory
head: cannot open 'IGNORE' for reading: No such file or directory


In [36]:
pd.read_json('./sample_data/anscombe.json')   --- IGNORE ---

SyntaxError: invalid syntax (3698854580.py, line 1)

## Other I/O functionalities
They are listed here https://pandas.pydata.org/pandas-docs/stable/user_guide/io.html and are suitable for: XML, HDF, Feather, Parquet files as well as for interaction with various databases and much more.

## Handling missing data and duplicates

In the real world it's often the case that there are missing or duplicate values in the dataset. They can be easily delt with in Pandas.

In [ ]:
df = pd.DataFrame([
  [0,       1     ],
  [2,       0     ],
  [2,       np.nan],
  [np.nan,  np.nan]
])

df

Incomplete observations can be filtered out.

In [ ]:
df.dropna()

In [ ]:
df.dropna(how='all')

It is also possible to "fill the gaps".

In [ ]:
df.fillna(0)

In [ ]:
df.fillna(0).drop_duplicates()

In [ ]:
df.fillna({0: 0, 1: 1})

## Data access

Let us observe how we can access and manipulate DataFrame structure.

In [42]:
from sklearn.datasets import fetch_california_housing
df = fetch_california_housing(as_frame=True).frame
df

MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
0      8.3252      41.0  6.984127   1.023810       322.0  2.555556     37.88   
1      8.3014      21.0  6.238137   0.971880      2401.0  2.109842     37.86   
2      7.2574      52.0  8.288136   1.073446       496.0  2.802260     37.85   
3      5.6431      52.0  5.817352   1.073059       558.0  2.547945     37.85   
4      3.8462      52.0  6.281853   1.081081       565.0  2.181467     37.85   
...       ...       ...       ...        ...         ...       ...       ...   
20635  1.5603      25.0  5.045455   1.133333       845.0  2.560606     39.48   
20636  2.5568      18.0  6.114035   1.315789       356.0  3.122807     39.49   
20637  1.7000      17.0  5.205543   1.120092      1007.0  2.325635     39.43   
20638  1.8672      18.0  5.329513   1.171920       741.0  2.123209     39.43   
20639  2.3886      16.0  5.254717   1.162264      1387.0  2.616981     39.37   

       Longitude  MedHouseVal  
0        -122.23        4.526  
1        -122.22        3.585  
2        -122.24        3.521  
3        -122.25        3.413  
4        -122.25        3.422  
...          ...          ...  
20635    -121.09        0.781  
20636    -121.21        0.771  
20637    -121.22        0.923  
20638    -121.32        0.847  
20639    -121.24        0.894  

[20640 rows x 9 columns]

### describe & info
Once the dataframe is loaded `describe` and `info`methods provide an easy way to obtain a high level view of the underlying data.

In [43]:
df.describe()

MedInc      HouseAge      AveRooms     AveBedrms    Population  \
count  20640.000000  20640.000000  20640.000000  20640.000000  20640.000000   
mean       3.870671     28.639486      5.429000      1.096675   1425.476744   
std        1.899822     12.585558      2.474173      0.473911   1132.462122   
min        0.499900      1.000000      0.846154      0.333333      3.000000   
25%        2.563400     18.000000      4.440716      1.006079    787.000000   
50%        3.534800     29.000000      5.229129      1.048780   1166.000000   
75%        4.743250     37.000000      6.052381      1.099526   1725.000000   
max       15.000100     52.000000    141.909091     34.066667  35682.000000   

           AveOccup      Latitude     Longitude   MedHouseVal  
count  20640.000000  20640.000000  20640.000000  20640.000000  
mean       3.070655     35.631861   -119.569704      2.068558  
std       10.386050      2.135952      2.003532      1.153956  
min        0.692308     32.540000   -124.350000      0.149990  
25%        2.429741     33.930000   -121.800000      1.196000  
50%        2.818116     34.260000   -118.490000      1.797000  
75%        3.282261     37.710000   -118.010000      2.647250  
max     1243.333333     41.950000   -114.310000      5.000010

In [44]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20640 entries, 0 to 20639
Data columns (total 9 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   MedInc       20640 non-null  float64
 1   HouseAge     20640 non-null  float64
 2   AveRooms     20640 non-null  float64
 3   AveBedrms    20640 non-null  float64
 4   Population   20640 non-null  float64
 5   AveOccup     20640 non-null  float64
 6   Latitude     20640 non-null  float64
 7   Longitude    20640 non-null  float64
 8   MedHouseVal  20640 non-null  float64
dtypes: float64(9)
memory usage: 1.4 MB


### head & tail
we can use these methods to look at the first and last entries

In [45]:
df.head()

MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
0  8.3252      41.0  6.984127   1.023810       322.0  2.555556     37.88   
1  8.3014      21.0  6.238137   0.971880      2401.0  2.109842     37.86   
2  7.2574      52.0  8.288136   1.073446       496.0  2.802260     37.85   
3  5.6431      52.0  5.817352   1.073059       558.0  2.547945     37.85   
4  3.8462      52.0  6.281853   1.081081       565.0  2.181467     37.85   

   Longitude  MedHouseVal  
0    -122.23        4.526  
1    -122.22        3.585  
2    -122.24        3.521  
3    -122.25        3.413  
4    -122.25        3.422

In [46]:
df.tail()

MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
20635  1.5603      25.0  5.045455   1.133333       845.0  2.560606     39.48   
20636  2.5568      18.0  6.114035   1.315789       356.0  3.122807     39.49   
20637  1.7000      17.0  5.205543   1.120092      1007.0  2.325635     39.43   
20638  1.8672      18.0  5.329513   1.171920       741.0  2.123209     39.43   
20639  2.3886      16.0  5.254717   1.162264      1387.0  2.616981     39.37   

       Longitude  MedHouseVal  
20635    -121.09        0.781  
20636    -121.21        0.771  
20637    -121.22        0.923  
20638    -121.32        0.847  
20639    -121.24        0.894

In [47]:
df.head(8)

MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
0  8.3252      41.0  6.984127   1.023810       322.0  2.555556     37.88   
1  8.3014      21.0  6.238137   0.971880      2401.0  2.109842     37.86   
2  7.2574      52.0  8.288136   1.073446       496.0  2.802260     37.85   
3  5.6431      52.0  5.817352   1.073059       558.0  2.547945     37.85   
4  3.8462      52.0  6.281853   1.081081       565.0  2.181467     37.85   
5  4.0368      52.0  4.761658   1.103627       413.0  2.139896     37.85   
6  3.6591      52.0  4.931907   0.951362      1094.0  2.128405     37.84   
7  3.1200      52.0  4.797527   1.061824      1157.0  1.788253     37.84   

   Longitude  MedHouseVal  
0    -122.23        4.526  
1    -122.22        3.585  
2    -122.24        3.521  
3    -122.25        3.413  
4    -122.25        3.422  
5    -122.25        2.697  
6    -122.25        2.992  
7    -122.25        2.414

### rename
can be used to rename column labels.

In [50]:
# df.rename(columns={'longitude': 'lon', 'latitude': 'lat'})
df

MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
0      8.3252      41.0  6.984127   1.023810       322.0  2.555556     37.88   
1      8.3014      21.0  6.238137   0.971880      2401.0  2.109842     37.86   
2      7.2574      52.0  8.288136   1.073446       496.0  2.802260     37.85   
3      5.6431      52.0  5.817352   1.073059       558.0  2.547945     37.85   
4      3.8462      52.0  6.281853   1.081081       565.0  2.181467     37.85   
...       ...       ...       ...        ...         ...       ...       ...   
20635  1.5603      25.0  5.045455   1.133333       845.0  2.560606     39.48   
20636  2.5568      18.0  6.114035   1.315789       356.0  3.122807     39.49   
20637  1.7000      17.0  5.205543   1.120092      1007.0  2.325635     39.43   
20638  1.8672      18.0  5.329513   1.171920       741.0  2.123209     39.43   
20639  2.3886      16.0  5.254717   1.162264      1387.0  2.616981     39.37   

       Longitude  MedHouseVal  
0        -122.23        4.526  
1        -122.22        3.585  
2        -122.24        3.521  
3        -122.25        3.413  
4        -122.25        3.422  
...          ...          ...  
20635    -121.09        0.781  
20636    -121.21        0.771  
20637    -121.22        0.923  
20638    -121.32        0.847  
20639    -121.24        0.894  

[20640 rows x 9 columns]

### Accessing a subset of data

To select multiple columns specify list of labels within square brackets.

In [51]:
df[['Longitude', 'Latitude']]

Longitude  Latitude
0        -122.23     37.88
1        -122.22     37.86
2        -122.24     37.85
3        -122.25     37.85
4        -122.25     37.85
...          ...       ...
20635    -121.09     39.48
20636    -121.21     39.49
20637    -121.22     39.43
20638    -121.32     39.43
20639    -121.24     39.37

[20640 rows x 2 columns]

Single column can be retrieved by specifying its name within square brackets. Note that this will yield a `Series` object which behaves a little different than DataFrames, but in general is a representation of single varaible values.

In [52]:
df['Longitude']

0       -122.23
1       -122.22
2       -122.24
3       -122.25
4       -122.25
          ...  
20635   -121.09
20636   -121.21
20637   -121.22
20638   -121.32
20639   -121.24
Name: Longitude, Length: 20640, dtype: float64

If column name does not contain spaces and other special character it can accesed as an attribute.

In [54]:
df.Longitude

0       -122.23
1       -122.22
2       -122.24
3       -122.25
4       -122.25
          ...  
20635   -121.09
20636   -121.21
20637   -121.22
20638   -121.32
20639   -121.24
Name: Longitude, Length: 20640, dtype: float64

Subset of a DataFrame can be selected using `loc` attribute. You need to specify index and column labels. Note that it supports slicing with respect to column names. Also, as [documentation](https://pandas.pydata.org/pandas-docs/version/0.15.2/indexing.html#different-choices-for-indexing) says *note that contrary to usual python slices, both the start and the stop are included!*.

In [55]:
df.loc[5:10, 'Population': 'median_house_value']

KeyError: 'median_house_value'

It also possible to select data with slices correspong to indices and columns **positions** (as in NumPy).

In [ ]:
df.iloc[5:10, -4:]

Boolean arrays (including boolean `Series` object) can be used as well.

In [ ]:
index = df.median_house_value < 25000
index

In [ ]:
df.loc[index, 'median_house_value']

It is also possible to do

In [ ]:
df[index]

## Merging dataframes
can be used to merge multiple arrays into one.


In [56]:
df_1 = pd.DataFrame(np.arange(9).reshape(3, 3), columns=['a', 'b', 'c'], index=[0, 1, 2])
df_2 = pd.DataFrame(np.arange(9).reshape(3, 3), columns=['b', 'c', 'd'], index=[1, 2, 3])

In [57]:
df_1

a  b  c
0  0  1  2
1  3  4  5
2  6  7  8

In [58]:
df_2

b  c  d
1  0  1  2
2  3  4  5
3  6  7  8

There are to basic methods for merging dataframes. The first one is `append` that concatenates dataframes along the index.

In [59]:
df_1 = pd.concat([df_1, df_2])

The second one is join that concatenates dataframes along the columns.

In [60]:
df_1.join(df_2, how='outer', lsuffix='_left', rsuffix='_right')

a  b_left  c_left  d_left  b_right  c_right  d_right
0  0.0       1       2     NaN      NaN      NaN      NaN
1  3.0       4       5     NaN      0.0      1.0      2.0
1  NaN       0       1     2.0      0.0      1.0      2.0
2  6.0       7       8     NaN      3.0      4.0      5.0
2  NaN       3       4     5.0      3.0      4.0      5.0
3  NaN       6       7     8.0      6.0      7.0      8.0

These two methods are special cases of more general function `concat` that allows more advanced oprations. There is also `merge` method which makes use of database-like joins.

All of these methods are described and explained here: https://pandas.pydata.org/pandas-docs/stable/user_guide/merging.html

## Operations on dataframes
Pandas can be used to perfrom mathematical and other operations on the underlying data.

### Elementwise operations
The most common use case is to create a new column being a result of some operation on existing columns.

In [64]:
print(df.head())
people_per_household = df.Population / df.MedHouseVal
people_per_household

   MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
0  8.3252      41.0  6.984127   1.023810       322.0  2.555556     37.88   
1  8.3014      21.0  6.238137   0.971880      2401.0  2.109842     37.86   
2  7.2574      52.0  8.288136   1.073446       496.0  2.802260     37.85   
3  5.6431      52.0  5.817352   1.073059       558.0  2.547945     37.85   
4  3.8462      52.0  6.281853   1.081081       565.0  2.181467     37.85   

   Longitude  MedHouseVal  
0    -122.23        4.526  
1    -122.22        3.585  
2    -122.24        3.521  
3    -122.25        3.413  
4    -122.25        3.422  


0          71.144498
1         669.735007
2         140.869071
3         163.492529
4         165.108124
            ...     
20635    1081.946223
20636     461.738003
20637    1091.007584
20638     874.852420
20639    1551.454139
Length: 20640, dtype: float64

In [ ]:
df['people_per_household'] = people_per_household
df

More general function can be applied with use of `apply` method

In [65]:
def func(row):
  if row.latitude > 0:
    ns_hemisphere = 'N'
  else:
    ns_hemisphere = 'S'

  if row.longitude > 0:
    ew_hemisphere = 'E'
  else:
    ew_hemisphere = 'W'

  gps_label = f"{abs(row.latitude)} {ns_hemisphere}, {abs(row.longitude)} {ew_hemisphere}"

  return gps_label

df.apply(func, axis=1).head()

AttributeError: 'Series' object has no attribute 'latitude'

### GroupBy
can be used to split a dataframe into multiple based on some condition/values

In [ ]:
values = df.median_house_value.round(-4)
values


In [ ]:
gb = df.groupby(values)
gb

Groups labels and indices can be accessed under `groups` attribute and themselves with the use of `get_group` method.

In [ ]:
gb.groups

In [ ]:
gb.get_group(10000)

The main use case of the `groupby` functionality is to split dataframe based on some condition, perform some aggregation within groups and then join the results. An example of such an operation is calcualtin mean within each group.

In [ ]:
gb.mean().head()

You can read more about `groupby` funcionality in teh pandas user guide: https://pandas.pydata.org/pandas-docs/stable/user_guide/groupby.html

## Pandas Exercises

### Pandas exercise: IMDB movies

In this exercise we will look into IMDB Top 1000 movies dataset. We will first load it into a dataframe.


In [75]:
url = "https://raw.githubusercontent.com/peetck/IMDB-Top1000-Movies/master/IMDB-Movie-Data.csv"
df = pd.read_csv(url)
df

Rank                    Title                     Genre  \
0       1  Guardians of the Galaxy   Action,Adventure,Sci-Fi   
1       2               Prometheus  Adventure,Mystery,Sci-Fi   
2       3                    Split           Horror,Thriller   
3       4                     Sing   Animation,Comedy,Family   
4       5            Suicide Squad  Action,Adventure,Fantasy   
..    ...                      ...                       ...   
995   996     Secret in Their Eyes       Crime,Drama,Mystery   
996   997          Hostel: Part II                    Horror   
997   998   Step Up 2: The Streets       Drama,Music,Romance   
998   999             Search Party          Adventure,Comedy   
999  1000               Nine Lives     Comedy,Family,Fantasy   

                                           Description              Director  \
0    A group of intergalactic criminals are forced ...            James Gunn   
1    Following clues to the origin of mankind, a te...          Ridley Scott   
2    Three girls are kidnapped by a man with a diag...    M. Night Shyamalan   
3    In a city of humanoid animals, a hustling thea...  Christophe Lourdelet   
4    A secret government agency recruits some of th...            David Ayer   
..                                                 ...                   ...   
995  A tight-knit team of rising investigators, alo...             Billy Ray   
996  Three American college students studying abroa...              Eli Roth   
997  Romantic sparks occur between two dance studen...            Jon M. Chu   
998  A pair of friends embark on a mission to reuni...        Scot Armstrong   
999  A stuffy businessman finds himself trapped ins...      Barry Sonnenfeld   

                                                Actors  Year  \
0    Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...  2014   
1    Noomi Rapace, Logan Marshall-Green, Michael Fa...  2012   
2    James McAvoy, Anya Taylor-Joy, Haley Lu Richar...  2016   
3    Matthew McConaughey,Reese Witherspoon, Seth Ma...  2016   
4    Will Smith, Jared Leto, Margot Robbie, Viola D...  2016   
..                                                 ...   ...   
995  Chiwetel Ejiofor, Nicole Kidman, Julia Roberts...  2015   
996  Lauren German, Heather Matarazzo, Bijou Philli...  2007   
997  Robert Hoffman, Briana Evigan, Cassie Ventura,...  2008   
998  Adam Pally, T.J. Miller, Thomas Middleditch,Sh...  2014   
999  Kevin Spacey, Jennifer Garner, Robbie Amell,Ch...  2016   

     Runtime (Minutes)  Rating   Votes  Revenue (Millions)  Metascore  
0                  121     8.1  757074              333.13         76  
1                  124     7.0  485820              126.46         65  
2                  117     7.3  157606              138.12         62  
3                  108     7.2   60545              270.32         59  
4                  123     6.2  393727              325.02         40  
..                 ...     ...     ...                 ...        ...  
995                111     6.2   27585                0.00         45  
996                 94     5.5   73152               17.54         46  
997                 98     6.2   70699               58.01         50  
998                 93     5.6    4881                0.00         22  
999                 87     5.3   12435               19.64         11  

[1000 rows x 12 columns]

Part 1.

1. How much should movie earn to be among 2% best earning movies?
2. Select movies that earned more than this value.
3. Calculate their mean rating.

In [86]:
# Make sure all columns are shown

pd.set_option("display.max_columns", None)
pd.set_option("display.max_colwidth", None)
pd.set_option("display.expand_frame_repr", False)

### YOUR CODE BEGINS HERE ###
# Get movies which are in top 2% best earning movies
df.sort_values(by='Revenue (Millions)', ascending=False, inplace=True)
top_2_percent_index = int(len(df) * 0.02)
top_2_percent_movies = df.iloc[:top_2_percent_index]
# print(top_2_percent_movies)
answer = top_2_percent_movies['Rating'].mean()
### YOUR CODE ENDS HERE ###

In [87]:
# Your code in cell(s) above
assert np.isclose(answer, 7.615)

Part 2.
1. Calculate mean metascore and mean revenue per director. (use groupby)
2. Restrict results to the directors that directed more than four movies. (look up `value_counts` method)
3. Sort the results with respect to the metascore and display the rating as a dataframe.
4. Who is the 3rd director it this rating and what is his/her mean revenue?

In [ ]:
### YOUR CODE BEGINS HERE ###

### YOUR CODE ENDS HERE ###

In [ ]:
print(f"Mean revenue of {director} is {revenue:.3f}.")
assert np.isclose(revenue, 43.242)

### Pandas exercises: Anscombe's quartet

In this exercise we will look into so called Anscombe's quartet that visualizes pitfalls of some statistics used to describe datasets

![image](https://upload.wikimedia.org/wikipedia/commons/e/ec/Anscombe%27s_quartet_3.svg)

Firstly, let us load the data.

In [88]:
df = pd.read_json("./sample_data/anscombe.json")
df

FileNotFoundError: File ./sample_data/anscombe.json does not exist

For each series calculate its
- mean value of $X$
- mean value of $Y$
- sample variance of $X$
- sample variance of $Y$
- covariance beetwen $X$ and $Y$

and verify whether for each series you get
- mean of $X$ = 9,
- var of $X$ = 11,
- mean of $Y$ = 7.5 +- 0.01,
- var of $Y$ = 4.125 +- 0.01,
- cov of $X$ and $Y$ = 5.50 +- 0.01,

(you can try to make use of `aggregate` function of a `GroupBy` object)

In [ ]:
### YOUR CODE BEGINS HERE ###

### YOUR CODE ENDS HERE ###

<center><img src='https://drive.google.com/uc?id=1BXZ0u3562N_MqCLcekI-Ens77Kk4LpPm'></center>